src/service/app_state.py

```diff
@@ -0,0 +1,125 @@
+"""
+Application state information and retrieval functions.
+
+All functions assume that the application state has been initialized via
+calling the build_app() method.
+"""
+
+import asyncio
+import logging
+import os
+from typing import NamedTuple, Optional
+
+from fastapi import FastAPI, Request
+
+from src.service.kb_auth import KBaseAuth, KBaseUser
+
+logger = logging.getLogger(__name__)
+
+
+class AppState(NamedTuple):
+    """Holds application state."""
+
+    auth: KBaseAuth
+
+
+class RequestState(NamedTuple):
+    """Holds request specific state."""
+
+    user: Optional[KBaseUser]
+
+
+async def build_app(app: FastAPI) -> None:
+    """
+    Build the application state.
+
+    Args:
+        app: The FastAPI app.
+    """
+    logger.info("Initializing application state...")
+
+    # Initialize auth with KBase auth URL and admin roles from environment variables
+    auth_url = os.environ.get("KBASE_AUTH_URL", "https://ci.kbase.us/services/auth/")
+    admin_roles = os.environ.get("KBASE_ADMIN_ROLES", "KBASE_ADMIN").split(",")
+
+    logger.info("Connecting to KBase auth service...")
+    auth = await KBaseAuth.create(auth_url, full_admin_roles=admin_roles)
+    logger.info("KBase auth service connected")
+
+    # Store components in app state
+    app.state._auth = auth
+    app.state._spark_state = AppState(auth=auth)
+    logger.info("Application state initialized")
+
+
+async def destroy_app_state(app: FastAPI) -> None:
+    """
+    Destroy the application state, shutting down services and releasing resources.
+
+    Args:
+        app: The FastAPI app.
+    """
+    # Currently no resources need to be explicitly cleaned up
+    # https://docs.aiohttp.org/en/stable/client_advanced.html#graceful-shutdown
+    await asyncio.sleep(0.250)
+    logger.info("Application state destroyed")
+
+
+def get_app_state(request: Request) -> AppState:
+    """
+    Get the application state from a request.
+
+    Args:
+        request: The FastAPI request.
+
+    Returns:
+        The application state.
+
+    Raises:
+        ValueError: If app state has not been initialized.
+    """
+    return _get_app_state_from_app(request.app)
+
+
+def _get_app_state_from_app(app: FastAPI) -> AppState:
+    """
+    Get the application state from a FastAPI app.
+
+    Args:
+        app: The FastAPI app.
+
+    Returns:
+        The application state.
+
+    Raises:
+        ValueError: If app state has not been initialized.
+    """
+    if not hasattr(app.state, "_spark_state") or not app.state._spark_state:
+        raise ValueError("App state has not been initialized")
+    return app.state._spark_state
+
+
+def set_request_user(request: Request, user: Optional[KBaseUser]) -> None:
```

Comment:
Are you using an older python version where you can't just `KbaseUser | None`?

Reply:
no. Updated to use `KBaseUser | None`.

Reply:
There's still a number of places in this pr that use Optional

Reply:
👍 I will get those fixed in my future PRs.